README.md
# Website Version Controller
PHP script that makes switching versions of webpage easy.  
The provided code allows you to decide which website version to choose simply by defining appropriate constant. I have provided you with template needed to write full system based on database data - only thing to do is to add code fetching data from your DB. Already written code checks for any missing files or inexistant versions.
# License
Code is provided with GNU GPLv3 license. Feel free to modify and redistribute the project!  
For full license text, go to LICENSE file
# Preparing VC to use
Code available on GitHub is already configured for a simple use scenario. You can simply copy all your website files to `versions` subdirectory. And voilà!
## Where to put my website files?
Your website files have to be copied to `versions/[version_name]/` directory. I have done my best to set all my code to be working transparently to you. Relative paths in `include` statements work as expected (`[version_name]` directory is considered as base path).
# Configuring VC
All configuration data is stored in `vc_user.php` file. Change settings by adjusting constants values.
## `DEFAULT_VERSION`
Default: `default`  
This constant stores default version to use. In default combination this field determines included version. It can be overwritten for a specific user by function `getVersionFromKeyPair()` (see below).
## `ERROR_ON_INEXISTANT_VERSION`
Default: `false`  
When the script tries to load a non-existing version of your site, it can print an error message (on `true`) or redirect to version `DEFAULT_VERSION` (on `false`).
## `VERSIONS_DIRECTORY`
Default: `versions/`  
Defines directory with alternative website versions. Value must end with slash (`/`).
## `ROOT_FILE`
Default: `index.php`  
Stores the main file's name. Main file is included by a script. Path from URL is available in `$_GET['path']` variable (for example when requesting `http://example.com/somefile?foo=bar`, your file will behave like when requesting `http://example.com/ROOT_FILE?path=somefile&foo=bar`).
## `SERIAL_KEY_COOKIE_NAME` and `RANDOM_KEY_COOKIE_NAME`
Default: `skey` and `rkey` respectively  
These two constants define names of cookies holding two 32-byte keys. The keypair can be used to identify particular clients. Serial key can be generated by your code. Delete function `getNextSerialKey()` (see below) from file `vc_user.php` to make it random.
## `STATUS_PAGE_ENABLED` and `STATUS_PAGE_URL`
Default: `false` and `vc_status` respectively  
With these constants you can enable simple status page that shows current configuration of the Version Controller. Visit page with URL stored in second constant to display that page (eg. `http://example.com/vc_status`) By default disabled.
## `DISABLE_SITE`
Default: `enable`  
With these settings you can disable whole site for all users or all except some group (for example for maintenance). Set `DISABLE_SITE` to `disable_all` if you want to completely disable site, or `disable_partial` if you want to give access only to group of users. They can access site by having special cookie (see below). 
## `DISABLE_PASS_COOKIE_NAME` and `DISABLE_PASS_COOKIE_VALUE`
Default: `dkey` and `1234` respectively  
They define a special cookie that gives access to *partially locked* website. Only users having `DISABLE_PASS_COOKIE_NAME` set to `DISABLE_PASS_COOKIE_VALUE` will be perimtted to view your site.
## `DISABLE_REASON`
Default: empty  
Content of `DISABLE_REASON` will show up under the *This website is disabled* header.
## `ENABLE_DEBUG`
Default: `false`  
When debug is enabled, there will be shown some data such as session keypair, loaded site version and file, contents of `$_GET` array under each HTML or PHP document.
## `USE_COOKIES`
Default: `true`  
Cookies are used to save session keypair. You can use the keys to identify your visitors. Disabling cookies disable session keys generation too.
# Additional functions
The additional functions aren't necessary for the script to work. You can even delete them. But they are supposed to be used as connector to database to store user-version pairs. For example you have 10 testers who should use `beta` version, but the rest of your users should see the `default` site.
## `string getVersionFromKeyPair(string[])`
Argument: Key pair (\[0\] => serial, \[1\] => random)  
Returns: Version name for those keys  
Here you can overwrite the `DEFAULT_VERSION` setting for each user separately.
## `string getNextSerialKey()`
Returns: 64 characters long string consisting of hex digits  
Generate serial key for a new session here. If deleted, serial key would be randomized.
## `void saveSessionKeyPair(string[])`
Argument: Key pair (\[0\] => serial, \[1\] => random)  
You can save the keypair to database (eg. to ensure that keys are unique).
## `bool isSessionKeyPairValid(string[])`
Argument: Key pair (\[0\] => serial, \[1\] => random)  
Returns: Whether the keypair is valid for use (eg. unique)
# Used libraries
* *Random Compatibility Library* by Paragon Initiative Enterprises (https://github.com/paragonie/random_compat/)
- - -
*Website Version Controller*  
Copyright 2018, Marcin Szwarc


CHANGELOG.md
# Version Controller Changelog
## 1.0
Published on: 2018-01-26
* Initial release


vc_user.php
<?php
/**
 * Website Version Controler
 * User-defined functions and settings
 * 
 * Version 1.0
 * Released 2018-01-26
 * 
 * This file is part of Website Version Controller.
 *
 * Website Version Controller is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Website Version Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License
 * along with Website Version Controller. If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (c) 2018, Marcin Szwarc
 * 
 ********************************************************************************
 *                                                                              *
 * In this file, you should write code that integrates Version Controller with  *
 * your server environment (like connections with database).                    *
 * Please fill these functions with your code and set constants to values that  *
 * fit your system the best.                                                    *
 *                                                                              *
 ********************************************************************************
 */

define(DEFAULT_VERSION, 'default');           // Default version directory
define(ERROR_ON_INEXISTANT_VERSION, true);   // Whether to show error when code tries to run non-existing version of site (if false, redirection to DEFAULT_VERSION will be made)
define(VERSIONS_DIRECTORY, 'versions/');      // Directory containing versions, must end with slash
define(ROOT_FILE, 'index.php');               // Main file for each version

define(SERIAL_KEY_COOKIE_NAME, 'skey');       // Defines name of the serial key cookie
define(RANDOM_KEY_COOKIE_NAME, 'rkey');       // Defines name of the random key cookie

define(STATUS_PAGE_ENABLED, false);           // Defines if status page is enabled
define(STATUS_PAGE_URL, 'vc_status');         // Defines URL for status page

define(DISABLE_SITE, 'enable');               // Set to disable_all to disable whole site; set to disable_partial to disable users without special cookie; Anything else means 'enable'
define(DISABLE_PASS_COOKIE_NAME, 'dkey');     // Defines name of the cookie which gives access to site with disable_partial state
define(DISABLE_PASS_COOKIE_VALUE, '1234');    // Defines value that grants access to site with disable_partial state
define(DISABLE_REASON, '');                   // Defines reason for disabling this website

define(ENABLE_DEBUG, false);                  // Displays debug information (for example keys from cookies and selected version) on bottom of each site
define(USE_COOKIES, true);                    // Defines whether to use cookies for making a session (if you disable cookies, no user-identifying data will be created)


// Returns website version for the key pair ([0] => serial, [1] => random)
function getVersionFromKeyPair($key_pair){
  return DEFAULT_VERSION;
}

// Returns next 32-byte serial key to use is sessions
// Delete this function to generate random key
// Return type: string of 64 hex chars
function getNextSerialKey(){
  return '0000000000000000000000000000000000000000000000000000000000000000';
}

// Saves session keys to database
function saveSessionKeyPair($key_pair){
  
}

// Checks if key pair is valid (eg. not expired)
// true = valid; false = invalid
function isSessionKeyPairValid($key_pair){
  return true;
}
?>

vc_status.php
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <style>
      td, th {padding:3px 7px;border-bottom:1px solid #aaa}
    </style>
  </head>
  <body>
    <h1>Version Controller status</h1>
    <table cellspacing="0">
    <tr><th>Setting</th><th>Value</th></tr>
    <?php
    echo('<tr><td>Default version</td><td>'.DEFAULT_VERSION.'</td></tr>');
    echo('<tr><td>Versions directory</td><td>'.VERSIONS_DIRECTORY.'</td></tr>');
    echo('<tr><td>Root file</td><td>'.ROOT_FILE.'</td></tr>');
    echo('<tr><td>Serial key cookie name</td><td>'.SERIAL_KEY_COOKIE_NAME.'</td></tr>');
    echo('<tr><td>Random key cookie name</td><td>'.RANDOM_KEY_COOKIE_NAME.'</td></tr>');
    echo('<tr><td>Disable site</td><td>'.DISABLE_SITE.'</td></tr>');
    echo('<tr><td>Disable pass cookie name</td><td>'.DISABLE_PASS_COOKIE_NAME.'</td></tr>');
    echo('<tr><td>Disable pass cookie value</td><td>'.DISABLE_PASS_COOKIE_VALUE.'</td></tr>');
    echo('<tr><td>Disable reason</td><td>'.DISABLE_REASON.'</td></tr>');
    echo('<tr><td>Enable debug</td><td>'.(ENABLE_DEBUG ? 'true' : 'false').'</td></tr>');
    ?>
    </table>
    <hr />
    Version Controller <?php echo(VC_VERSION) ?> by Marcin Szwarc
  </body>
</html>

version_controller.php
<?php
/**
 * Website Version Controler
 * for easy version management
 *
 * Version 1.0
 * Released 2018-01-26
 *
 * This file is part of Website Version Controller.
 *
 * Website Version Controller is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Website Version Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License
 * along with Website Version Controller. If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (c) 2018, Marcin Szwarc
 *
 */

include('random-compat/random.php');          // Random number generator for PHP 5; can be omitted in PHP 7
                                              // Random Compatibility Library by Paragon Initiative Enterprises

require('vc_user.php');                       // User-defined functions
require('mime.php');                          // MIME type recognition

define('VC_VERSION', '1.0');                  // Stores script version

// Check whether site should be disabled
if(DISABLE_SITE == 'disable_all'){
  echo('<h1>This website is disabled.</h1>');
  echo(DISABLE_REASON);
  exit;
}
if(USE_COOKIES){
  if(DISABLE_SITE == 'disable_partial' && !isset($_COOKIE[DISABLE_PASS_COOKIE_NAME])){
    echo('<h1>This website is disabled.</h1>');
    echo(DISABLE_REASON);
    exit;
  }else{
    if(DISABLE_SITE == 'disable_partial' && $_COOKIE[DISABLE_PASS_COOKIE_NAME] != DISABLE_PASS_COOKIE_VALUE){
      echo('<h1>This website is disabled.</h1>');
      echo(DISABLE_REASON);
      exit;
    }
  }
}

// Check whether versions directory exist
if(!file_exists(VERSIONS_DIRECTORY)){
  echo("<h1>Version directory <code>'".VERSIONS_DIRECTORY."'</code> doesn't exist</h1>");
  echo('Please check your system and settings and try again.');
  echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
  exit;
}

// Check whether default version directory exist
if(!file_exists(VERSIONS_DIRECTORY.DEFAULT_VERSION)){
  echo("<h1>Default version <code>'".DEFAULT_VERSION."'</code> doesn't exist</h1>");
  echo('Please check your system and settings and try again.');
  echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
  exit;
}

// Check whether default version's root file exist
if(!file_exists(VERSIONS_DIRECTORY.DEFAULT_VERSION.'/'.ROOT_FILE)){
  echo("<h1>Root file <code>'".ROOT_FILE."'</code> for the default version <code>'".DEFAULT_VERSION."'</code> doesn't exist</h1>");
  echo('Please check your system and settings and try again.');
  echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
  exit;
}

// Show status page if requested
if(STATUS_PAGE_ENABLED == true && STATUS_PAGE_URL == $_GET['path']){
  include('vc_status.php');
  exit;
}

// Include root file of specified version
if(USE_COOKIES) $key_pair = readSessionKeyPair();
else $key_pair = array('0000000000000000000000000000000000000000000000000000000000000000', '0000000000000000000000000000000000000000000000000000000000000000');
if(function_exists('getVersionFromKeyPair')) $version = getVersionFromKeyPair($key_pair);
else $version = DEFAULT_VERSION;
if(!file_exists(VERSIONS_DIRECTORY.$version)){
  if(ERROR_ON_INEXISTANT_VERSION){
    echo("<h1>Requested version <code>'".$version."'</code> doesn't exist</h1>");
    echo('Please check your system and settings and try again.');
    echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
    exit;
  }else $version = DEFAULT_VERSION;
}
if(!file_exists(VERSIONS_DIRECTORY.$version.'/'.ROOT_FILE)){
  if(ERROR_ON_INEXISTANT_VERSION){
    echo("<h1>Requested version <code>'".$version."'</code> doesn't have a root file <code>'".ROOT_FILE."'</code></h1>");
    echo('Please check your system and settings and try again.');
    echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
    exit;
  }else $version = DEFAULT_VERSION;
}

if(isset($_GET['path'])) $get_path = $_GET['path']; else $get_path = '';

// Include code files, flush assets (PNGs, CSSs, JSs, etc.)
if(isAssetFile($get_path)){
  $path = VERSIONS_DIRECTORY.$version.'/'.$get_path;
  if(file_exists($path)) header('Content-Type: '.getMimeTypeByFileName($get_path));
  else{
    http_response_code(404);
    echo("<h1>404 Not Found</h1>");
    echo('Requested file <code>'.$get_path.'</code> wasn\'t found.');
    echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
    exit;
  }
}else{
  $path = VERSIONS_DIRECTORY.$version.'/'.ROOT_FILE;
}

define('CURRENT_VERSION', $version);
define('SERIAL_KEY', $keypair[0]);
define('RANDOM_KEY', $keypair[1]);
chdir(realpath(VERSIONS_DIRECTORY.$version));

include($path);



// Debug info
if(ENABLE_DEBUG){
  if(!isAssetFile($get_path)){
    echo('<h3>Version Controller debug info</h3>');
    echo('<code>Serial key: <b>'.$key_pair[0].'</b></code><br />');
    echo('<code>Random key: <b>'.$key_pair[1].'</b></code><br />');
    echo('<code>Loaded version: <b>'.$version.'</b></code><br />');
    echo('<code>Loaded file: <b>'.$path.'</b></code><br />');
    echo('<br /><code>$_GET array:</code><pre style="margin:0 2ch">');
    print_r($_GET);
    echo('</pre>');
    
    echo('<hr /><i>Version Controller '.VC_VERSION.'</i> by Marcin Szwarc');
  }
}




// Generates new key pair
// Returns array of two keys ([0] => serial, [1] => random)
function generateSessionKeyPair(){
  if(function_exists('getNextSerialKey')) $skey = getNextSerialKey();
  else $skey = getRandomKey();
  $rkey = getRandomKey();
  setcookie(SERIAL_KEY_COOKIE_NAME, $skey, 0);
  setcookie(RANDOM_KEY_COOKIE_NAME, $rkey, 0);
  $key_pair = array($skey, $rkey);
  if(function_exists('saveSessionKeyPair')) saveSessionKeyPair($key_pair);
  return $key_pair;
}

// Returns random 32-byte key
// Returns string of 64 hex chars
function getRandomKey(){
  if(!function_exists('random_bytes')) return randomStr(64);
  try{
    $string = bin2hex(random_bytes(32));
  }catch(Exception $e){
    $string = randomStr(64);
  }
  return $string;
}

// Reads keys from cookies or creates a new pair if keys do not exist or are invalid
// Returns array of session keys
function readSessionKeyPair(){
  if(isset($_COOKIE[SERIAL_KEY_COOKIE_NAME]) && isset($_COOKIE[RANDOM_KEY_COOKIE_NAME])){
    $skey = $_COOKIE[SERIAL_KEY_COOKIE_NAME];
    $rkey = $_COOKIE[RANDOM_KEY_COOKIE_NAME];
    $key_pair = array($skey, $rkey);
    if(!function_exists('isSessionKeyPairValid')) return $key_pair;
    if(isSessionKeyPairValid($key_pair)) return $key_pair;
    return generateSessionKeyPair();
  }else{
    return generateSessionKeyPair();
  }
}

// Fallback function for generating random key
// Returns pseudo-random string of hex chars
function randomStr($length){
  $str = '';
  $keyspace = '0123456789abcdef';
  $max = mb_strlen($keyspace, '8bit') - 1;
  for($i = 0; $i < $length; $i++){
    $str .= $keyspace[rand(0, $max)];
  }
  return $str;
}

// Check if file is an asset file (like graphics, CSS files or javascript files)
function isAssetFile($path){
  if(strpos($path, '.') === false) return false;
  $ext = substr($path, strrpos($path, '.') + 1);
  $non_asset = array('php', 'php3', 'php4', 'php5', 'phtml', 'html', 'inc');
  return !in_array($ext, $non_asset);
}
?>